# FUNCTION gradstat


The gradstat function extracts some relevant statistics from the image gradient computed in the [gradimg](gradimg.ipynb) function.

<ul> <strong> g = gradstat(f, mask[]) </strong>
       <ul> <li> <strong> Output </strong> </li>
           <ul> <li><strong>g: </strong> gradient statistics</li>
           </ul>
       </ul>
       <ul> <li> <strong> Input </strong> </li>
           <ul> <li><strong>f: </strong> ndarray, input image (2D or 3D)</li>
                <li><strong>mask: </strong> define the mask </li>
           </ul>
       </ul>
</ul>



In [2]:
import nbimporter
from auxiliary_functions import *
from scipy.stats import *
import numpy as np

def gradstat(f,mask=[]):
    grad = gradimg(f)
    b = iasecross(1)
    
    if (len(f.shape) == 3):                                                     # 3D Case
        if mask != []:
            for i in arange(mask.shape[0]):
                mask_aux = np.zeros(mask.shape[1]+2, mask.shape[2]+2)
                mask_aux[1:-1,1:-1] = mask[i]                                  # zero bounder
                mask[i] = iaero(mask_aux,b)[1:-1,1:-1]                         # extract the outside border
                #mask[i] = mask[i] ^ iacontour(mask[i])
            grad = grad[mask>0]
        else:
            m = grad>-1 # all pixels
            for i in np.arange(m.shape[0]):
                m_aux = np.zeros((m.shape[1]+2,m.shape[2]+2))
                m_aux[1:-1,1:-1] = m[i]                                        # zero bounder
                m[i] = iaero(m_aux,b)[1:-1,1:-1]                               # extract the outside border
                m = g>-1 
                for i in np.arange(m.shape[0]):
                    m[i] = m[i] ^ iacontour(m[i]) # extract the outside border
                    grad = g[m]
            grad = grad[mask>0]
    else:                                                                      # 2D Case
        if mask!=[]:
            mask_aux = np.zeros((mask.shape[0]+2,mask.shape[1]+2))
            mask_aux[1:-1,1:-1] = mask                                         # zero bounder
            mask = iaero(mask_aux,b)[1:-1,1:-1]                                # extract the outside border
            #mask = mask ^ iacontour(mask)
            grad = grad[mask>0]
        else:
            m = grad>-1                                                        # all pixels
            m_aux = np.zeros((m.shape[0]+2,m.shape[1]+2))
            m_aux[1:-1,1:-1] = m                                               # zero bounder
            m = iaero(m_aux,b)[1:-1,1:-1]                                      # extract the outside border
            #m = m ^ iacontour(m)
            grad = grad[m>0]
    M = np.size(grad)
    percNonZeros = 1.0*np.count_nonzero(grad)/M
    
    
    # stats = histstat(grad, np.bincount(grad.ravel()))
    return grad,[np.mean(np.ravel(grad)), np.var(np.ravel(grad)), skew(np.ravel(grad)), kurtosis(np.ravel(grad)), percNonZeros]

# EXAMPLES

Numerical example:

In [3]:
import nbimporter
import numpy as np
from gradimg import *
from auxiliary_functions import *

np.set_printoptions(precision=2)

f = np.array( [[2,2,0,1,1,1,0,0,0],
               [1,2,1,1,1,1,1,1,1],
               [1,1,1,1,1,1,1,1,1],
               [0,0,2,2,2,1,1,0,0]], dtype=np.uint8)

_,stats = gradstat(f)
print 'mean absolute gradient',stats[0]
print 'variance absolute gradient',stats[1]
print 'skewness absolute gradient',stats[2]
print 'kurtosis absolute gradient',stats[3]
print 'percentage of pixels with nonzero gradient',stats[4]

mean absolute gradient 0.743872397312
variance absolute gradient 0.375225285088
skewness absolute gradient 0.106509983523
kurtosis absolute gradient -0.842523037749
percentage of pixels with nonzero gradient 0.642857142857


Numerical example:

In [4]:
import nbimporter
import numpy as np
from gradimg import *
from auxiliary_functions import *

f = np.array( [[0,0,0,1,1,1,0,0,0],
               [1,1,1,1,1,1,1,1,1],
               [1,1,1,1,1,1,1,1,1],
               [0,0,0,1,1,1,1,0,0]], dtype=np.uint8)

grad,stats = gradstat(f,f>0)
print 'mean absolute gradient',stats[0]
print 'variance absolute gradient',stats[1]
print 'skewness absolute gradient',stats[2]
print 'kurtosis absolute gradient',stats[3]
print 'percentage of pixels with nonzero gradient',stats[4]

print 
grad, stats = gradstat(f)
print 'mean absolute gradient',stats[0]
print 'variance absolute gradient',stats[1]
print 'skewness absolute gradient',stats[2]
print 'kurtosis absolute gradient',stats[3]
print 'percentage of pixels with nonzero gradient',stats[4]

mean absolute gradient 0.0
variance absolute gradient 0.0
skewness absolute gradient 0.0
kurtosis absolute gradient -3.0
percentage of pixels with nonzero gradient 0.0

mean absolute gradient 0.5
variance absolute gradient 0.25
skewness absolute gradient 0.0
kurtosis absolute gradient -2.0
percentage of pixels with nonzero gradient 0.5


/home/lburini/Documents/IC/lib/python2.7/site-packages/ipykernel_launcher.py:30: DeprecationWarning: elementwise != comparison failed; this will raise an error in the future.


Numerical example 2:

In [8]:
import nbimporter 
from histstat import *
from histogram import *

f = np.array( [[0,0,0,1,1],
               [1,1,1,1,1],
               [1,1,1,1,1],
               [0,0,0,1,1],
               [1,1,1,1,1]], dtype=np.uint8)
g = gradimg(f)
h = hist(f)
print histstat(f, h)
print
print 'gradient[f]=',g
grad, stats = gradstat(f)
print 'mean absolute gradient',stats[0]
print 'variance absolute gradient',stats[1]
print 'skewness absolute gradient',stats[2]
print 'kurtosis absolute gradient',stats[3]
print 'percentage of pixels with nonzero gradient',stats[4]

mean absolute gradient 0.666666666667
variance absolute gradient 0.222222222222
skewness absolute gradient -0.707106781187
kurtosis absolute gradient -1.5
percentage of pixels with nonzero gradient 0.666666666667




**3D Example**

In [14]:
f = np.array( [[[0,0,0,1,1],
                [1,1,1,1,1],
                [1,1,1,1,1],
                [0,0,0,1,1]],
                [[0,0,0,1,1],
                 [1,1,1,1,1],
                 [1,1,1,1,1],
                 [0,0,0,1,1]]], dtype=np.uint8)
g = gradimg(f)
grad, stats = gradstat(f)
m = g>-1 # all pixels

print 'mean absolute gradient',stats[0]
print 'variance absolute gradient',stats[1]
print 'skewness absolute gradient',stats[2]
print 'kurtosis absolute gradient',stats[3]
print 'percentage of pixels with nonzero gradient',stats[4]

mean absolute gradient 1.27614237492
variance absolute gradient 0.038127305612
skewness absolute gradient -0.707106781187
kurtosis absolute gradient -1.5
percentage of pixels with nonzero gradient 1.0


# EQUATIONS

Let $G(i,j)$ be the gradient matrix computed with 3x3 neighbor

<strong>Mean absolute gradient</strong>
$$ GrMean = \frac{1}{M} \sum\limits_{i, j  \in  ROI} G(i,j)$$

<strong>Variance of absolute gradient</strong>
$$ GrVariance = \frac{1}{M} \sum\limits_{i, j  \in  ROI} {(G(i,j)- GrMean)^2}$$ 

<strong>Skewness of absolute gradient</strong>
$$ GrSkewness = \frac{1}{GrVariance^{1.5}} \frac{1}{M} \sum\limits_{i, j  \in  ROI} {(G(i,j)- GrMean)^3}$$ 

<strong>Kurtosis of absolute gradient</strong>
$$ GrSKurtosis = \frac{1}{GrVariance^2}\frac{1}{M} \sum\limits_{i, j  \in  ROI} (G(i,j) - GrMean)^4 - 3$$